cmd/cainjector/app/start.go

```diff
@@ -120,47 +108,52 @@ func (o InjectorControllerOptions) runCertificateBasedInjector(stopCh <-chan str
 		LeaderElectionID:        "cert-manager-cainjector-leader-election",
 		MetricsBindAddress:      "0",
 	})
-
 	if err != nil {
-		o.log.Error(err, "error creating manager")
-		os.Exit(1)
+		return fmt.Errorf("error creating manager: %v", err)
 	}
 
-	// TODO(directxman12): enabled controllers for separate injectors?
-	if err := cainjector.RegisterCertificateBased(mgr); err != nil {
-		o.log.Error(err, "error registering controllers")
-		os.Exit(1)
-	}
+	g, gctx := errgroup.WithContext(ctx)
 
-	if err := mgr.Start(stopCh); err != nil {
-		o.log.Error(err, "error running manager")
-		os.Exit(1)
-	}
-}
+	g.Go(func() (err error) {
+		defer func() {
+			o.log.Error(err, "manager goroutine exited")
+		}()
 
-func (o InjectorControllerOptions) runSecretBasedInjector(stopCh <-chan struct{}) {
-	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
-		Scheme:                  api.Scheme,
-		Namespace:               o.Namespace,
-		LeaderElection:          o.LeaderElect,
-		LeaderElectionNamespace: o.LeaderElectionNamespace,
-		LeaderElectionID:        "cert-manager-cainjector-leader-election-core",
-		MetricsBindAddress:      "0",
+		if err = mgr.Start(gctx.Done()); err != nil {
+			return fmt.Errorf("error running manager: %v", err)
+		}
+		return nil
 	})
 
-	if err != nil {
-		o.log.Error(err, "error creating core-only manager")
-		os.Exit(1)
-	}
+	<-mgr.Elected()
```

Comment:
This avoids running any of the caches or controllers unless / until this process is the elected leader.

Reply:
Maybe check for `gctx.Done()` here too, to avoid launching go routines if the context has already been cancelled.
I don't think errgroup does that.